addons/udes_stock/tests/test_batch_pickings.py

```diff
@@ -322,3 +322,114 @@ def test04_fix(self):
 
         picks.mapped("move_lines").action_refactor()
         self.assertFactorised()
+
+
+class TestCancelInBatch(BaseUDES):
+    """Tests for cancelling picks in a batch."""
+
+    @classmethod
+    def setUpClass(cls):
+        """Setup for test class."""
+        super().setUpClass()
+        PickingBatch = cls.env["stock.picking.batch"]
+        cls.batch = PickingBatch.create({"name": "batch001",
+                                         "user_id": cls.env.uid})
+        cls.create_quant(cls.apple.id, cls.test_location_01.id, 50)
+        cls.create_quant(cls.banana.id, cls.test_location_01.id, 50)
+        cls.pick1 = cls.create_picking(
+            picking_type=cls.picking_type_pick,
+            products_info=[{'product': cls.apple, 'qty': 2},
+                           {'product': cls.banana, 'qty': 4},
+                           ])
+        cls.pick2 = cls.create_picking(
+            picking_type=cls.picking_type_pick,
+            products_info=[{'product': cls.apple, 'qty': 6},
+                           {'product': cls.banana, 'qty': 8},
+                           ])
+        cls.pick3 = cls.create_picking(
+            picking_type=cls.picking_type_pick,
+            products_info=[{'product': cls.apple, 'qty': 10},
+                           {'product': cls.banana, 'qty': 12},
+                           ])
+        cls.picks = cls.pick1 | cls.pick2 | cls.pick3
+        cls.picks.action_confirm()
+
+    def test01_cancel_in_batch(self):
+        """Verify that picks can be cancelled in a batch
+
+        Verify that a batch in state ready allows a pick to be cancelled, and
+        that it's reserved quants (and moves) are freed accordingly
+
+        """
+        # Search the quants that we know will be reserved
+        Quant = self.env['stock.quant']
+        apples = Quant.search([('location_id', '=', self.test_location_01.id),
+                               ('product_id', '=', self.apple.id)])
+        bananas = Quant.search([('location_id', '=', self.test_location_01.id),
+                                ('product_id', '=', self.banana.id)])
+
+        # Put test picks in a batch and confirm them
+        picks = self.pick1 | self.pick2 | self.pick3
+        picks.write({'batch_id': self.batch.id})
+        self.batch.confirm_picking()
+
+        # Confirm batch state
+        self.assertEqual(self.batch.state, 'in_progress')
+
+        # Confirm reserved stock
+        reserved = {}
+        for move in self.pick1.move_line_ids:
+            reserved[move.product_id.name] = move.product_qty
+        self.assertEqual(reserved, {self.banana.name: 4, self.apple.name: 2})
```

Comment:
This block of code is repeated several few times.

The generation of reserved could be done on one line using list comprehension.

``reserved = {move.product_id.name:move.product_qty for move in self.pick1.move_line_ids}``

Asserting the reserved amount for a pick could be a method of the test class, e.g.

```
def assertReserved(self, pick, expected):
    reserved = {move.product_id.name:move.product_qty for move in pick.move_line_ids}
    self.assertEqual(reserved, expected)
```

It would be less code and more readable IMHO

Reply:
good suggestion, tyvm